File: sprint8/By lessons/lesson6code1/lesson6code1.kt
val N = s.length
// Создадим массив π, состоящий из N нулей.
val π = IntArray(N) { 0 }

for (i in 1 until N) {
    // i -- длина подстроки-префикса.
    val substring = s.substring(0, i)
    // Проверяем, перекрывается ли подстрока substring с собой по k символам.
    for (k in (i - 1) downTo 0) {
        // Для этого сравним префикс и суффикс соответствующих длин.
        val prefix = substring.substring(0, k)
        val suffix = substring.substring(i - k, i)
        if (prefix == suffix) {
            π[i - 1] = k  // Запишем значение π-функции.
            break         // Дальше не проверяем -- k идёт в порядке уменьшения.
        }
    }
}

File: sprint8/By lessons/lesson6code1/lesson6code1.swift
let N = s.count
// Создадим массив π, состоящий из N нулей.
var π = [Int](repeating: 0, count: N)

for i in 1..<N {
    // i -- длина подстроки-префикса.
    let substring = String(s.prefix(i))
    // Проверяем, перекрывается ли подстрока substring с собой по k символам.
    for k in stride(from: i - 1, through: 0, by: -1) {
        // Для этого сравним префикс и суффикс соответствующих длин.
        let prefix = String(substring.prefix(k))
        let suffix = String(substring.suffix(i - k))
        if prefix == suffix {
            // Запишем значение π-функции.
            π[i-1] = k
            // Дальше не проверяем -- k идёт в порядке уменьшения.
            break
        }
    }
}
